### PR TITLE
Narrow types with while statement

### LINKED ISSUE #270: While statement should narrow types
Specified here https://github.com/ballerina-platform/ballerina-spec/issues/840

Follow on from #129

### PR DESCRIPTION
**Pull request by manuranga**
Fixes #270

### CODE REVIEW

File: compiler/testSuite/12-narrow/while02-v.bal [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+                     // @output 2
+                     // @output 1
+      if count > 1 {
+         count = count - 1;
+      }
+      else {
+         count = false;
+      }
+   }
+   // boolean _ = count;
```
Comment (manuranga) [inline compiler/testSuite/12-narrow/while02-v.bal]: @jclark Shouldn't `count` be narrowed to `false` here ?

**Review by jclark — CHANGES_REQUESTED** [review]
This is including #831 which I don't think is ready yet.

File: compiler/testSuite/12-narrow/while02-v.bal (hunk 1 above)
Comment (jclark) [reply compiler/testSuite/12-narrow/while02-v.bal]: I don't think the spec says that yet.

Please add a spec issue with this example and subject "Normal completion of while-stmt should be able to cause variable type narrowing".

If you don't have a `break`, then after the `while` completes normally the loop condition while be false.

Comment (manuranga) [reply compiler/testSuite/12-narrow/while02-v.bal]: crated https://github.com/ballerina-platform/ballerina-spec/issues/1019

**Comment by manuranga** [comment]
Please review #839 first.

**Review by jclark — CHANGES_REQUESTED** [review]
Can you rebase this? Looks like it is straightforward given #839.

**Pull request merged**